src/inmanta/server/protocol.py

```diff
@@ -173,75 +177,114 @@ def _get_slice_sequence(self):
             This prevents database connection from being closed too early. This order in which the endpoint
             are started, is hardcoded in the get_server_slices() method in server/bootloader.py
         """
-        await super(Server, self).stop()
         if not self.running:
             return
-
         self.running = False
-        LOGGER.debug("Stopping Server Rest Endpoint")
-        await self._transport.stop()
-        for endpoint in reversed(list(self._get_slice_sequence())):
-            LOGGER.debug("Stopping %s", endpoint)
-            await endpoint.stop()
 
-        await self._transport.join()
-        await self.disconnect_database()
+        await super(Server, self).stop()
 
-    async def connect_database(self) -> None:
-        """ Connect to the database
-        """
-        database_host = opt.db_host.get()
-        database_port = opt.db_port.get()
+        order = list(reversed(self._get_slice_sequence()))
 
-        database_username = opt.db_username.get()
-        database_password = opt.db_password.get()
-        await data.connect(database_host, database_port, opt.db_name.get(), database_username, database_password)
-        LOGGER.info("Connected to PostgreSQL database %s on %s:%d", opt.db_name.get(), database_host, database_port)
+        for endpoint in order:
+            LOGGER.debug("Pre Stopping %s", endpoint.name)
+            await endpoint.prestop()
 
-    async def disconnect_database(self) -> None:
-        """ Disconnect the database
-        """
-        await data.disconnect()
+        for endpoint in order:
+            LOGGER.debug("Stopping %s", endpoint.name)
+            await endpoint.stop()
 
 
 class ServerSlice(inmanta.protocol.endpoints.CallTarget, TaskHandler):
     """
-        An API serving part of the server.
+        Base class for server extensions offering zero or more api endpoints
+
+        Extensions developers should override the lifecycle methods:
+
+        * :func:`ServerSlice.prestart`
+        * :func:`ServerSlice.start`
+        * :func:`ServerSlice.prestop`
+        * :func:`ServerSlice.stop`
+        * :func:`ServerSlice.get_dependencies`
+
+        To register endpoints that server static content, either use :func:'add_static_handler' or :func:'add_static_content'
+        To create endpoints, use the annotation based mechanism
+
+        To schedule recurring tasks, use self._sched
+        To schedule background tasks, use self.add_background_task
     """
 
     def __init__(self, name: str) -> None:
         super().__init__()
 
         self._name: str = name
         self._handlers: List[routing.Rule] = []
-        self._sched = Scheduler("server slice")  # FIXME: why has each slice its own scheduler?
+        self._sched = Scheduler(f"server slice {name}")  # FIXME: why has each slice its own scheduler?
```

Comment:
I think we can remove this comment. Background tasks are also per instance, however there it is with inheritance and scheduler with delegation.